src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp

```diff
@@ -932,9 +1002,385 @@ namespace
             const float jacobian = abs(cos_ih) * square(eta / sqrt_denom);
             return jacobian * mdf.pdf(wo, h, alpha_x, alpha_y, gamma);
         }
+
+        static float ec_lobe_ratio(
+            const float eta,
+            const float favg,
+            const float favg_rcp_eta,
+            const float eavg,
+            const float eavg_rcp_eta,
+            const float reflection_weight,
+            const float refraction_weight)
+        {
+            const float a = (1.0f - favg) / (1.0f - eavg_rcp_eta);
+            const float b = (1.0f - favg_rcp_eta) * square(eta) / (1.0f - eavg);
+            const float ratio = b / (a + b);
+
+            const float r_ratio = reflection_weight * ratio;
+            const float t_ratio = refraction_weight * (1.0f - ratio);
+            const float sum_ratios = r_ratio + t_ratio;
+
+            if (sum_ratios != 0.0f)
+                return r_ratio / sum_ratios;
+
+            return 0.0f;
+        }
+
+        template <typename MDF>
+        static void add_energy_compensation_term(
+            const MDF&                  mdf,
+            const InputValues*          values,
+            const Vector3f&             normal,
+            const Vector3f&             outgoing,
+            const Vector3f&             incoming,
+            Spectrum&                   value)
+        {
+            if (values->m_energy_compensation != 0.0f)
+            {
+                const float cos_in = dot(incoming, normal);
+                const float cos_on = dot(outgoing, normal);
+
+                const float roughness = values->m_roughness;
+
+                const float eta = values->m_precomputed.m_eta;
+                const float rcp_eta = 1.0f / values->m_precomputed.m_eta;
```

Comment:
Are we not precomputing 1/eta already? If not, would it make sense?

Reply:
We weren't precomputing it before. 
We are computing 1.0 / eta 2 times when energy compensation is on.